File: code/backend/python/app/adapters/auth/jwt_adapter.py
import jwt
import time
import uuid
from app.core.ports.auth.jwt_port import JwtPort
from app.core.models.auth.token_info import TokenInfo

class JwtAdapter(JwtPort):
    def __init__(self, secret: str, access_token_minutes: int, refresh_token_days: int):
        self.secret = secret
        self.access_token_minutes = access_token_minutes
        self.refresh_token_days = refresh_token_days

    def generate_access_token(self, user_uuid: str, username: str, role: str) -> str:
        payload = {
            "sub": user_uuid,
            "username": username,
            "role": role,
            "type": "access",
            "iat": int(time.time()),
            "exp": int(time.time()) + (self.access_token_minutes * 60),
            "jti": str(uuid.uuid4())
        }
        return jwt.encode(payload, self.secret, algorithm="HS256")

    def generate_refresh_token(self, user_uuid: str) -> str:
        payload = {
            "sub": user_uuid,
            "type": "refresh",
            "iat": int(time.time()),
            "exp": int(time.time()) + (self.refresh_token_days * 24 * 60 * 60),
            "jti": str(uuid.uuid4())
        }
        return jwt.encode(payload, self.secret, algorithm="HS256")

    def get_access_token_expiration_ms(self) -> int:
        return (int(time.time()) + (self.access_token_minutes * 60)) * 1000

    def get_refresh_token_expiration_ms(self) -> int:
        return (int(time.time()) + (self.refresh_token_days * 24 * 60 * 60)) * 1000

    def parse_token(self, token: str) -> TokenInfo:
        try:
            payload = jwt.decode(token, self.secret, algorithms=["HS256"])
            return TokenInfo(
                sub=payload["sub"],
                username=payload.get("username", ""),
                role=payload.get("role", "PLAYER"),
                type=payload["type"],
                iat=payload["iat"],
                exp=payload["exp"],
                jti=payload.get("jti")
            )
        except Exception as e:
            raise ValueError(f"TOKEN_PARSE_ERROR: {str(e)}")

    def validate_token(self, token: str) -> bool:
        try:
            jwt.decode(token, self.secret, algorithms=["HS256"])
            return True
        except Exception:
            return False

File: code/backend/php/src/Adapter/Auth/JwtAdapter.php
<?php

namespace Games\Paths\Adapter\Auth;

use Games\Paths\Core\Port\Auth\JwtPort;
use Games\Paths\Core\Domain\Auth\TokenInfo;
use Firebase\JWT\JWT;
use Firebase\JWT\Key;
use Ramsey\Uuid\Uuid;

/**
 * JWT Adapter — replicates the Python JwtAdapter logic exactly.
 * Uses HS256 with the same secret, same payload structure, same expiry calculation.
 */
class JwtAdapter implements JwtPort
{
    private string $secret;
    private int $accessTokenMinutes;
    private int $refreshTokenDays;

    public function __construct(string $secret, int $accessTokenMinutes = 30, int $refreshTokenDays = 7)
    {
        $this->secret = $secret;
        $this->accessTokenMinutes = $accessTokenMinutes;
        $this->refreshTokenDays = $refreshTokenDays;
    }

    public function generateAccessToken(string $userUuid, string $username, string $role): string
    {
        $now = time();
        $payload = [
            'sub'      => $userUuid,
            'username' => $username,
            'role'     => $role,
            'type'     => 'access',
            'iat'      => $now,
            'exp'      => $now + ($this->accessTokenMinutes * 60),
            'jti'      => Uuid::uuid4()->toString(),
        ];
        return JWT::encode($payload, $this->secret, 'HS256');
    }

    public function generateRefreshToken(string $userUuid): string
    {
        $now = time();
        $payload = [
            'sub'  => $userUuid,
            'type' => 'refresh',
            'iat'  => $now,
            'exp'  => $now + ($this->refreshTokenDays * 24 * 60 * 60),
            'jti'  => Uuid::uuid4()->toString(),
        ];
        return JWT::encode($payload, $this->secret, 'HS256');
    }

    public function getAccessTokenExpirationMs(): int
    {
        return (time() + ($this->accessTokenMinutes * 60)) * 1000;
    }

    public function getRefreshTokenExpirationMs(): int
    {
        return (time() + ($this->refreshTokenDays * 24 * 60 * 60)) * 1000;
    }

    public function parseToken(string $token): TokenInfo
    {
        try {
            $payload = JWT::decode($token, new Key($this->secret, 'HS256'));
            return new TokenInfo(
                $payload->sub,
                $payload->username ?? '',
                $payload->role ?? 'PLAYER',
                $payload->type,
                $payload->iat,
                $payload->exp,
                $payload->jti ?? null
            );
        } catch (\Exception $e) {
            throw new \RuntimeException("TOKEN_PARSE_ERROR: " . $e->getMessage(), 401);
        }
    }

    public function validateToken(string $token): bool
    {
        try {
            JWT::decode($token, new Key($this->secret, 'HS256'));
            return true;
        } catch (\Exception $e) {
            return false;
        }
    }
}
